// @orca-so/whirlpool-client-sdk/dist/context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhirlpoolContext = void 0;
const anchor_1 = require("@project-serum/anchor");
const whirlpool_json_1 = __importDefault(require("./artifacts/whirlpool.json"));
class WhirlpoolContext {
    constructor(provider, program, opts) {
        this.connection = provider.connection;
        this.wallet = provider.wallet;
        this.opts = opts;
        // It's a hack but it works on Anchor workspace *shrug*
        this.program = program;
        this.provider = provider;
    }
    static from(connection, wallet, programId, opts = anchor_1.Provider.defaultOptions()) {
        const provider = new anchor_1.Provider(connection, wallet, opts);
        const program = new anchor_1.Program(whirlpool_json_1.default, programId, provider);
        return new WhirlpoolContext(provider, program, opts);
    }
    static fromWorkspace(provider, program, opts = anchor_1.Provider.defaultOptions()) {
        return new WhirlpoolContext(provider, program, opts);
    }
    static withProvider(provider, programId, opts = anchor_1.Provider.defaultOptions()) {
        const program = new anchor_1.Program(whirlpool_json_1.default, programId, provider);
        return new WhirlpoolContext(provider, program, opts);
    }
}
exports.WhirlpoolContext = WhirlpoolContext;
