// @saberhq/token-utils/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.u64 = exports.TOKEN_PROGRAM_ID = exports.SPLToken = exports.NATIVE_MINT = exports.ASSOCIATED_TOKEN_PROGRAM_ID = exports.ZERO = exports.validateU256 = exports.validateU64 = exports.TEN = exports.Rounding = exports.Percent = exports.parseBigintIsh = exports.ONE = exports.MAX_U256 = exports.MAX_U64 = exports.makeDecimalMultiplier = exports.Fraction = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./ata"), exports);
tslib_1.__exportStar(require("./instructions"), exports);
tslib_1.__exportStar(require("./layout"), exports);
tslib_1.__exportStar(require("./price"), exports);
tslib_1.__exportStar(require("./splTokenRegistry"), exports);
tslib_1.__exportStar(require("./token"), exports);
tslib_1.__exportStar(require("./tokenAmount"), exports);
tslib_1.__exportStar(require("./tokenList"), exports);
tslib_1.__exportStar(require("./tokenOwner"), exports);
tslib_1.__exportStar(require("./tokenProvider"), exports);
// re-export token-math types
// so consumers don't need to use them
var token_math_1 = require("@ubeswap/token-math");
Object.defineProperty(exports, "Fraction", { enumerable: true, get: function () { return token_math_1.Fraction; } });
Object.defineProperty(exports, "makeDecimalMultiplier", { enumerable: true, get: function () { return token_math_1.makeDecimalMultiplier; } });
Object.defineProperty(exports, "MAX_U64", { enumerable: true, get: function () { return token_math_1.MAX_U64; } });
Object.defineProperty(exports, "MAX_U256", { enumerable: true, get: function () { return token_math_1.MAX_U256; } });
Object.defineProperty(exports, "ONE", { enumerable: true, get: function () { return token_math_1.ONE; } });
Object.defineProperty(exports, "parseBigintIsh", { enumerable: true, get: function () { return token_math_1.parseBigintIsh; } });
Object.defineProperty(exports, "Percent", { enumerable: true, get: function () { return token_math_1.Percent; } });
Object.defineProperty(exports, "Rounding", { enumerable: true, get: function () { return token_math_1.Rounding; } });
Object.defineProperty(exports, "TEN", { enumerable: true, get: function () { return token_math_1.TEN; } });
Object.defineProperty(exports, "validateU64", { enumerable: true, get: function () { return token_math_1.validateU64; } });
Object.defineProperty(exports, "validateU256", { enumerable: true, get: function () { return token_math_1.validateU256; } });
Object.defineProperty(exports, "ZERO", { enumerable: true, get: function () { return token_math_1.ZERO; } });
// re-export SPL token types
tslib_1.__exportStar(require("./common"), exports);
var spl_token_1 = require("@solana/spl-token");
Object.defineProperty(exports, "ASSOCIATED_TOKEN_PROGRAM_ID", { enumerable: true, get: function () { return spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID; } });
Object.defineProperty(exports, "NATIVE_MINT", { enumerable: true, get: function () { return spl_token_1.NATIVE_MINT; } });
Object.defineProperty(exports, "SPLToken", { enumerable: true, get: function () { return spl_token_1.Token; } });
Object.defineProperty(exports, "TOKEN_PROGRAM_ID", { enumerable: true, get: function () { return spl_token_1.TOKEN_PROGRAM_ID; } });
Object.defineProperty(exports, "u64", { enumerable: true, get: function () { return spl_token_1.u64; } });
