// @orca-so/whirlpool-client-sdk/dist/instructions/increase-liquidity-ix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildIncreaseLiquidityIx = void 0;
const spl_token_1 = require("@solana/spl-token");
function buildIncreaseLiquidityIx(context, params) {
    const { liquidityAmount, tokenMaxA, tokenMaxB, whirlpool, positionAuthority, position, positionTokenAccount, tokenOwnerAccountA, tokenOwnerAccountB, tokenVaultA, tokenVaultB, tickArrayLower, tickArrayUpper, } = params;
    const ix = context.program.instruction.increaseLiquidity(liquidityAmount, tokenMaxA, tokenMaxB, {
        accounts: {
            whirlpool,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            positionAuthority,
            position,
            positionTokenAccount,
            tokenOwnerAccountA,
            tokenOwnerAccountB,
            tokenVaultA,
            tokenVaultB,
            tickArrayLower,
            tickArrayUpper,
        },
    });
    return {
        instructions: [ix],
        cleanupInstructions: [],
        signers: [],
    };
}
exports.buildIncreaseLiquidityIx = buildIncreaseLiquidityIx;
