// @orca-so/whirlpool-client-sdk/dist/instructions/initialize-pool-ix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildInitPoolIx = void 0;
const web3_js_1 = require("@solana/web3.js");
const spl_token_1 = require("@solana/spl-token");
function buildInitPoolIx(context, params) {
    const program = context.program;
    const { initSqrtPrice, tokenMintA, tokenMintB, whirlpoolConfigKey, whirlpoolPda, feeTierKey, tokenVaultAKeypair, tokenVaultBKeypair, tickSpacing, funder, } = params;
    const whirlpoolBumps = {
        whirlpoolBump: whirlpoolPda.bump,
    };
    const ix = program.instruction.initializePool(whirlpoolBumps, tickSpacing, initSqrtPrice, {
        accounts: {
            whirlpoolsConfig: whirlpoolConfigKey,
            tokenMintA: tokenMintA,
            tokenMintB: tokenMintB,
            funder,
            whirlpool: whirlpoolPda.publicKey,
            tokenVaultA: tokenVaultAKeypair.publicKey,
            tokenVaultB: tokenVaultBKeypair.publicKey,
            feeTier: feeTierKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
    });
    return {
        instructions: [ix],
        cleanupInstructions: [],
        signers: [tokenVaultAKeypair, tokenVaultBKeypair],
    };
}
exports.buildInitPoolIx = buildInitPoolIx;
