// @saberhq/token-utils/dist/esm/tokenAmount.js
import { u64 } from "@solana/spl-token";
import { parseBigintIsh, TokenAmount as UTokenAmount, validateU64, } from "@ubeswap/token-math";
import BN from "bn.js";
export class TokenAmount extends UTokenAmount {
    // amount _must_ be raw, i.e. in the native representation
    constructor(token, amount) {
        super(token, amount);
        validateU64(this.raw);
    }
    /**
     * Parses a token amount from a decimal representation.
     * @param token
     * @param uiAmount
     * @returns
     */
    static parse(token, uiAmount) {
        const prev = UTokenAmount.parseFromString(token, uiAmount);
        return new TokenAmount(token, prev.raw);
    }
    add(other) {
        const result = super.add(other);
        return new TokenAmount(this.token, result.raw);
    }
    subtract(other) {
        const result = super.subtract(other);
        return new TokenAmount(this.token, result.raw);
    }
    multiplyBy(percent) {
        const result = super.multiplyBy(percent);
        return new TokenAmount(this.token, result.raw);
    }
    reduceBy(percent) {
        const result = super.reduceBy(percent);
        return new TokenAmount(this.token, result.raw);
    }
    /**
     * Divides this TokenAmount by a raw integer.
     * @param other
     * @returns
     */
    divideByInteger(other) {
        return new TokenAmount(this.token, this.toU64().div(new BN(parseBigintIsh(other).toString())));
    }
    /**
     * Formats the token amount with units and decimal adjustment, e.g. "100.42 SOL"
     * @returns
     */
    formatUnits(format = { groupSeparator: "," }) {
        return `${this.toExact(format)} ${this.token.symbol}`;
    }
    /**
     * String representation of this token amount.
     */
    toString() {
        return `TokenAmount[Token=(${this.token.toString()}), amount=${this.toExact()}`;
    }
    /**
     * JSON representation of the token amount.
     */
    toJSON() {
        return {
            _isTA: true,
            mint: this.token.address,
            uiAmount: this.toExact(),
        };
    }
    /**
     * Converts this to the raw u64 used by the SPL library
     * @returns
     */
    toU64() {
        return new u64(this.raw.toString());
    }
}
